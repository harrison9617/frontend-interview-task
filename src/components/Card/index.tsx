import { Allowance } from '@/types/allowances'
import React from 'react'
import StatusBar from '../StatusBar'
import useCalculatePercent from '@/hooks/useCalculatePercent'

const Card = ({
  id,
  name,
  type,
  currency,
  amount,
  spent,
  renewal,
  active,
}: Allowance) => {
  const percent = useCalculatePercent(spent, amount)

  return (
    <div
      id={id}
      className="border-[1px] rounded-lg flex flex-col justify-between gap-6"
    >
      <div className="px-5 pt-6">
        <p className="text-md font-medium text-gray-500 mb-3">{name}</p>
        <p className="text-sm text-gray-400">
          {type === 'card' && 'Spend Card'}
          {type === 'expense' && 'Expense'}
        </p>
      </div>
      {active ? (
        <div className="px-5 py-4">
          <StatusBar
            percent={percent}
            text={`
              ${Intl.NumberFormat('en-GB', {
                style: 'currency',
                currency: currency,
              }).format(amount)} / ${renewal}
            `}
          />
        </div>
      ) : (
        <div
          className="p-4 bg-gray-100 cursor-pointer"
          onClick={(id) =>
            console.log(
              `TODO: Send an UPDATE request to API to change allowance with an id of ${id} active status to ${!active}`
            )
          }
        >
          <p className="text-sm font-medium text-lime-500">Activate {type}</p>
        </div>
      )}
    </div>
  )
}

export default Card
