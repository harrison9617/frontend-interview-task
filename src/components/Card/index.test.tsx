import { render, waitFor } from '@testing-library/react'
import Card from '@/components/Card'
import '@testing-library/jest-dom'
import 'jest-fetch-mock'

describe('Card', () => {
  it('renders as expected when currency is GBP', async () => {
    const { container } = render(
      <Card
        id={'1'}
        name="Name"
        type="card"
        currency="GBP"
        amount={100}
        spent={50}
        renewal="week"
        active={false}
      />
    )
    expect(container).toMatchSnapshot()
  })
  it('renders as expected when currency is USD', async () => {
    const { container } = render(
      <Card
        id={'1'}
        name="Name"
        type="card"
        currency="USD"
        amount={100}
        spent={50}
        renewal="week"
        active={false}
      />
    )
    expect(container).toMatchSnapshot()
  })
  it('renders as expected when currency is EUR', async () => {
    const { container } = render(
      <Card
        id={'1'}
        name="Name"
        type="card"
        currency="EUR"
        amount={100}
        spent={50}
        renewal="week"
        active={false}
      />
    )
    expect(container).toMatchSnapshot()
  })
  it('renders card type correctly by displaying Spend Card', async () => {
    const { getByText } = render(
      <Card
        id={'1'}
        name="Name"
        type="card"
        currency="GBP"
        amount={100}
        spent={50}
        renewal="week"
        active={false}
      />
    )
    expect(getByText('Spend Card')).toBeInTheDocument()
  })
  it('renders expense type correctly by displaying Expense', async () => {
    const { getByText } = render(
      <Card
        id={'1'}
        name="Name"
        type="expense"
        currency="GBP"
        amount={100}
        spent={50}
        renewal="week"
        active={false}
      />
    )
    expect(getByText('Expense')).toBeInTheDocument()
  })
})
