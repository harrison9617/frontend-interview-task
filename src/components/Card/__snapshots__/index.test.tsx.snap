// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders as expected when currency is EUR 1`] = `
<div>
  <div
    class="border-[1px] rounded-lg flex flex-col justify-between gap-6"
    id="1"
  >
    <div
      class="px-5 pt-6"
    >
      <p
        class="text-md font-medium text-gray-500 mb-3"
      >
        Name
      </p>
      <p
        class="text-sm text-gray-400"
      >
        Spend Card
      </p>
    </div>
    <div
      class="p-4 bg-gray-100 cursor-pointer"
    >
      <p
        class="text-sm font-medium text-lime-500"
      >
        Activate 
        card
      </p>
    </div>
  </div>
</div>
`;

exports[`Card renders as expected when currency is GBP 1`] = `
<div>
  <div
    class="border-[1px] rounded-lg flex flex-col justify-between gap-6"
    id="1"
  >
    <div
      class="px-5 pt-6"
    >
      <p
        class="text-md font-medium text-gray-500 mb-3"
      >
        Name
      </p>
      <p
        class="text-sm text-gray-400"
      >
        Spend Card
      </p>
    </div>
    <div
      class="p-4 bg-gray-100 cursor-pointer"
    >
      <p
        class="text-sm font-medium text-lime-500"
      >
        Activate 
        card
      </p>
    </div>
  </div>
</div>
`;

exports[`Card renders as expected when currency is USD 1`] = `
<div>
  <div
    class="border-[1px] rounded-lg flex flex-col justify-between gap-6"
    id="1"
  >
    <div
      class="px-5 pt-6"
    >
      <p
        class="text-md font-medium text-gray-500 mb-3"
      >
        Name
      </p>
      <p
        class="text-sm text-gray-400"
      >
        Spend Card
      </p>
    </div>
    <div
      class="p-4 bg-gray-100 cursor-pointer"
    >
      <p
        class="text-sm font-medium text-lime-500"
      >
        Activate 
        card
      </p>
    </div>
  </div>
</div>
`;
